### PR TITLE
Update to Babylon 7 (fixes #481)

### LINKED ISSUE #481: Update to use Babylon 7 (alpha)
Ref https://github.com/babel/babel-eslint/pull/463

This will help us lint the newer proposals only on Babel 7

- [ ] Comment below you are going to do this (for others to know) - we can have multiple people reviewing
- [ ] Read [CONTRIBUTING.md](https://github.com/babel/babel-eslint/blob/master/CONTRIBUTING.md) (make sure you have yarn, and are on at least node 4)
- [ ] Join [slack community](https://slack.babeljs.io/) for more immediate help (if people are available)
- [ ] Update the babel packages to 7.0.0-alpha.12 https://github.com/babel/babel-eslint/blob/54ed517d2ecb3afb93ac338c019f8dff224918c2/package.json#L15-L18
- [ ] make tests pass (most likely the node types are different)
  - [ ] if it makes the diffs easier to read we can switch to jest too
- [ ] possibly add new test for new syntax (I can help with this when we get to it)

### PR DESCRIPTION
**Pull request by soda0289**
> EDIT by @hzoo, Fixes #481

I did not add any tests. What new syntax in babylon 7 is there to test?

### CODE REVIEW

**Comment by hzoo** [comment]
Awesome to see that's all that's required for the tests that are there

File: package.json [hunk 1]
```diff
@@ -12,10 +12,10 @@
     "url": "https://github.com/babel/babel-eslint.git"
   },
   "dependencies": {
-    "babel-code-frame": "^6.22.0",
-    "babel-traverse": "^6.23.1",
-    "babel-types": "^6.23.0",
-    "babylon": "^6.17.0"
+    "babel-code-frame": "^7.0.0-alpha.12",
```
Comment (hzoo) [inline package.json]: I would pin the versions while we're in alpha (so `^` should be removed)

**Comment by hzoo** [comment]
https://github.com/babel/babylon#plugins

- `numericSeparator`, `optionalChaining`, `importMeta`, `classPrivateProperties`, `dynamicImport`

Although we can just deal with it when we get issues for it as well. We may need to make changes to https://github.com/babel/eslint-plugin-babel to handle new syntax.

**Comment by soda0289** [comment]
I pinned the depencies for babel to 7.0.0.alpha12 and babylon to 7.0.0.beta13

File: package.json [hunk 2]
```diff
@@ -12,10 +12,10 @@
     "url": "https://github.com/babel/babel-eslint.git"
   },
   "dependencies": {
-    "babel-code-frame": "^6.22.0",
-    "babel-traverse": "^6.23.1",
-    "babel-types": "^6.23.0",
-    "babylon": "^6.17.0"
+    "babel-code-frame": "7.0.0-alpha.12",
```
Comment (hzoo) [inline package.json]: There's a deprecation for code-frame, https://github.com/babel/babel/pull/5808/files

https://github.com/babel/babel-eslint/blob/54ed517d2ecb3afb93ac338c019f8dff224918c2/index.js#L421

**Comment by hzoo** [comment]
Also can you add https://github.com/soda0289/babel-eslint/blob/481868d404aa606946e68b6504a6f0bdf8a744d9/index.js#L385-L400 to the plugins array the new ones I mentioned above? https://github.com/babel/babylon/#plugins

**Comment by hzoo** [comment]
- [ ] need to handle no-undef/no-unused-vars for private properties

```js
export class C {
  #d = 1; // error  'd' is not defined  no-undef
}
```

**Comment by soda0289** [comment]
@hzoo I can write some tests for those cases.

**Comment by soda0289** [comment]
Update the eslint-scope monkey patch to handle private class properties and included tests. Anything else that needs tests/updating?

**Comment by hzoo** [comment]
Just testing it on a codebase it is good until we find other errors (in theory we should just against all eslint rules but it's hard to setup)

Tried and gave up a few times to setup https://github.com/babel/babel-eslint/issues/62

**Comment by hzoo** [comment]
Awesome work!!

**Comment by nodkz** [comment]
It's quite strange but `optionalChaning` does not work for me with babel-eslint@8.0.0-alpha.13

screen shot 2017-07-13 at 21 47 58

https://twitter.com/nodkz/status/885529331212193793

Using such `.babelrc.js`: https://gist.github.com/nodkz/41e189ff22325a27fe6a5ca81df2cb91
Code transpiles and executes perfectly.

Maybe I miss some tricky option?

**Pull request merged**